# Grandeur

<img src="https://www.roadtripryan.com/go/resources/content/utah/wasatch/grandeur-peak/user-submitted/ryancornia-1505057017043.jpg" width="500" align="left" />

Named after the beautiful [Grandeur Peak](https://www.alltrails.com/trail/us/utah/grandeur-peak-east-trail-from-church-fork)

Image Credit: [ryancornia](https://www.roadtripryan.com/go/resources/content/utah/wasatch/grandeur-peak/user-submitted/ryancornia-1505057017043.jpg)

Location:  40.707, -111.76, 8,299 ft (2,421 m) summit

More information about the trail leading up to this landmark can be found at [https://utah.com/hiking/grandeur-peak](https://utah.com/hiking/grandeur-peak)

Grandeur is a [Nextflow](https://www.nextflow.io/) workflow developed by [@erinyoung](https://github.com/erinyoung) at the [Utah Public Health Laborotory](https://uphl.utah.gov/). "Grandeur" is intended to be a species agnostic sequencing analysis workflow to paired-end Illumina sequencing quality control and assurance (QC) and serotyping in a local public health laboratory. 

"Grandeur" is meant to augment CDC's [PHOENIX](https://github.com/CDCgov/phoenix) nextflow workflow, which is the official recommended usage. In principle, the contigs generated by PHOENIX undergo additional quality metric and serotyping steps, with a heavy emphasis on [fastANI](https://github.com/ParBLiSS/FastANI) and [AMRFinderPlus](https://github.com/ncbi/amr).

"Grandeur" can also be a standalone workflow that takes paired-end Illumina reads, removes adaptors with [fastp](https://github.com/OpenGene/fastp) and PHIX with [bbduk](https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbduk-guide/), and creates contigs through _de novo_ alignment of the reads with [spades](https://cab.spbu.ru/software/spades/).

"Grandeur" is also a workflow of the [staphb-toolkit](https://github.com/StaPH-B/staphb_toolkit)

## Dependencies

- [Nextflow](https://www.nextflow.io/docs/latest/getstarted.html)
- [Singularity](https://singularity.lbl.gov/install-linux) or [Docker](https://docs.docker.com/get-docker/)
- [Git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)

## Usage

Default workflow that takes fastq files, runs them through QC/serotyping/etc, creates contig files
```
# using singularity
nextflow run UPHL-BioNGS/Grandeur -profile singularity --reads <path to reads>
# using docker
nextflow run UPHL-BioNGS/Grandeur -profile docker --fastas <path to fastas>
```

WARNING : "Grandeur" will automatically grab any fastq files in `workflow.launchDir + '/reads'`, fasta files in `workflow.launchDir + '/fastas'`, and gff files in `workflow.launchDir + '/gff'`. This is a **feature** of the workflow.

## Wiki sections

The README got too long, so it's been moved to a [wiki](https://github.com/UPHL-BioNGS/Grandeur/wiki). There are several covered topics including:
- [Installation](https://github.com/UPHL-BioNGS/Grandeur/wiki/Installation)
- [Usage](https://github.com/UPHL-BioNGS/Grandeur/wiki/USAGE)
  - [Standalone](https://github.com/UPHL-BioNGS/Grandeur/wiki/USAGE#standalone)
  - [After PHOENIX or another workflow](https://github.com/UPHL-BioNGS/Grandeur/wiki/USAGE#after-phoenix)
  - [Phylogenetic or Relatedness Analysis](https://github.com/UPHL-BioNGS/Grandeur/wiki/Phylogenetic-Analysis)
  - [Using config Files](https://github.com/UPHL-BioNGS/Grandeur/wiki/config-file)
  - [Examples](https://github.com/UPHL-BioNGS/Grandeur/wiki/examples)
- [Subworkflow explanations](https://github.com/UPHL-BioNGS/Grandeur/wiki/subworkflows)
- [User supplied reference files and databases (optional)](https://github.com/UPHL-BioNGS/Grandeur/wiki/User-supplied-reference-files-and-databases)
- [FAQ](https://github.com/UPHL-BioNGS/Grandeur/wiki/FAQ)

## Acknowledgements

Grandeur wouldn't be possible without the following tools:
- [amrfinderplus](https://www.ncbi.nlm.nih.gov/pathogens/antimicrobial-resistance/AMRFinder/) - identification of genes associated with antimicrobial resistence
- [bbduk](https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbduk-guide/) - removal of PhiX
- [bbmap](https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbduk-guide/) - maps reads to contigs for blobtools
- [blastn](https://www.ncbi.nlm.nih.gov/books/NBK279684/) - read identification with blobtools
- [blobtools](https://blobtools.readme.io/docs) - contamination
- [datasets](https://github.com/ncbi/datasets) - downloads genomes from NCBI
- [fastani](https://github.com/ParBLiSS/FastANI) - species evaluator
- [fastp](https://github.com/OpenGene/fastp) - cleaning reads
- [fastqc](https://github.com/s-andrews/FastQC) - fastq file QC
- [fastq-scan](https://github.com/rpetit3/fastq-scan) - quality metrics for fastq files in json format
- [iqtree2](http://www.iqtree.org/) - phylogenetic tree creation - used after core genome alignment
- [kleborate](https://github.com/katholt/Kleborate) - Klebsiella serotyping
- [kraken2](https://ccb.jhu.edu/software/kraken2/) - contamination
- [mash](https://github.com/marbl/Mash) - species identifier
- [mlst](https://github.com/tseemann/mlst) - identification of MLST subtype
- [multiqc](https://multiqc.info/) - summarizes QC efforts
- [plasmidfinder](https://bitbucket.org/genomicepidemiology/plasmidfinder/) - MLST typing for plasmids
- [prokka](https://github.com/tseemann/prokka) - gene annotation - used for core genome alignment
- [quast](http://quast.sourceforge.net/quast) - contig QC
- [roary](https://sanger-pathogens.github.io/Roary/) - core genome alignment - optional
- [seqsero2](https://github.com/denglab/SeqSero2) - Salmonella serotyping
- [serotypefinder](https://cge.cbs.dtu.dk/services/SerotypeFinder/) - E. coli serotyping
- [shigatyper](https://github.com/CFSAN-Biostatistics/shigatyper) - Shigella serotyping
- [snp-dists](https://github.com/tseemann/snp-dists) - SNP matrix - used after core genome aligment
- [spades](https://cab.spbu.ru/software/spades/) - _de novo_ alignment
- [sra toolkit](https://hpc.nih.gov/apps/sratoolkit.html) - for downloading fastq files for testing
